fix bernoulli_ for functionalization by making it a primitive wrt torch_dispatch

bdhirsh opened the pull request:
Context: `bernoulli_.float` is currently a primitive op, but it's functional counterpart `bernoulli.p` is `CompositeImplicitAutograd`.

This is a problem when we're using functionalization on top of python tracing: functionalization will convert `bernoulli_.float` into `bernoulli.p`, but the decomposition for `bernoulli.p` will run before hitting python, and happily convert back into `bernoulli_.float`.

This doesn't seems to come up very often, but I think the right fix is to just make `bernoulli.p` a primitive w.r.t. the Python key too, by making it `CompositeExplicitAutograd`.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #85844
* #85875
* #85681
* #82601
* #82602
* __->__ #82771
* #83265

bdhirsh committed:
fix bernoulli_ for functionalization by making it a primitive wrt torch_dispatch

[ghstack-poisoned]

bdhirsh committed:
Update on "fix bernoulli_ for functionalization by making it a primitive wrt torch_dispatch"

Context: `bernoulli_.float` is currently a primitive op, but it's functional counterpart `bernoulli.p` is `CompositeImplicitAutograd`.

This is a problem when we're using functionalization on top of python tracing: functionalization will convert `bernoulli_.float` into `bernoulli.p`, but the decomposition for `bernoulli.p` will run before hitting python, and happily convert back into `bernoulli_.float`.

This doesn't seems to come up very often, but I think the right fix is to just make `bernoulli.p` a primitive w.r.t. the Python key too, by making it `CompositeExplicitAutograd`.




[ghstack-poisoned]

ezyang reviewed (approved):
okey dokey

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -943,6 +943,12 @@
+    # This op needs to be a primitive w.r.t. Python,
+    # because functionalization wants to convert bernoulli_.float -> bernoulli.p.
+    # If we leave it as CompositeImplicitAutograd, then we'll run this kernel
+    # If we leave it as CompositeImplicitAutograd, then we'll run this kernel
```
duplicated line in comment?

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -333,6 +333,10 @@
   p: zeros_like(p)
   result: self_t.zero_()
 
+- name: bernoulli.p(Tensor self, float p, *, Generator? generator=None) -> Tensor
+  self: zeros_like(grad)
+  result: auto_element_wise
```
The behavior today is that bernoulli.p returns a Tensor without requires_grad. Does this PR change the behavior by giving it a backwards formula?
```
In [6]: x = torch.randn(3, 3, requires_grad=True)

In [7]: y = torch.bernoulli(x, 0.5)

In [8]: y
Out[8]:
tensor([[1., 1., 0.],
        [0., 1., 0.],
        [1., 0., 0.]])

```

albanD replied:
Correct, please mark `output_differentiability = [False]` here to preserve that behavior

bdhirsh replied:
whoops, thanks :)

bdhirsh committed:
Update on "fix bernoulli_ for functionalization by making it a primitive wrt torch_dispatch"

Context: `bernoulli_.float` is currently a primitive op, but it's functional counterpart `bernoulli.p` is `CompositeImplicitAutograd`.

This is a problem when we're using functionalization on top of python tracing: functionalization will convert `bernoulli_.float` into `bernoulli.p`, but the decomposition for `bernoulli.p` will run before hitting python, and happily convert back into `bernoulli_.float`.

This doesn't seems to come up very often, but I think the right fix is to just make `bernoulli.p` a primitive w.r.t. the Python key too, by making it `CompositeExplicitAutograd`.




[ghstack-poisoned]

albanD reviewed (approved):
Duplicate comments needs to be removed. But good otherwise.

bdhirsh committed:
Update on "fix bernoulli_ for functionalization by making it a primitive wrt torch_dispatch"

Context: `bernoulli_.float` is currently a primitive op, but it's functional counterpart `bernoulli.p` is `CompositeImplicitAutograd`.

This is a problem when we're using functionalization on top of python tracing: functionalization will convert `bernoulli_.float` into `bernoulli.p`, but the decomposition for `bernoulli.p` will run before hitting python, and happily convert back into `bernoulli_.float`.

This doesn't seems to come up very often, but I think the right fix is to just make `bernoulli.p` a primitive w.r.t. the Python key too, by making it `CompositeExplicitAutograd`.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix bernoulli_ for functionalization by making it a primitive wrt torch_dispatch"

Context: `bernoulli_.float` is currently a primitive op, but it's functional counterpart `bernoulli.p` is `CompositeImplicitAutograd`.

This is a problem when we're using functionalization on top of python tracing: functionalization will convert `bernoulli_.float` into `bernoulli.p`, but the decomposition for `bernoulli.p` will run before hitting python, and happily convert back into `bernoulli_.float`.

This doesn't seems to come up very often, but I think the right fix is to just make `bernoulli.p` a primitive w.r.t. the Python key too, by making it `CompositeExplicitAutograd`.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix bernoulli_ for functionalization by making it a primitive wrt torch_dispatch"

Context: `bernoulli_.float` is currently a primitive op, but it's functional counterpart `bernoulli.p` is `CompositeImplicitAutograd`.

This is a problem when we're using functionalization on top of python tracing: functionalization will convert `bernoulli_.float` into `bernoulli.p`, but the decomposition for `bernoulli.p` will run before hitting python, and happily convert back into `bernoulli_.float`.

This doesn't seems to come up very often, but I think the right fix is to just make `bernoulli.p` a primitive w.r.t. the Python key too, by making it `CompositeExplicitAutograd`.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix bernoulli_ for functionalization by making it a primitive wrt torch_dispatch"

Context: `bernoulli_.float` is currently a primitive op, but it's functional counterpart `bernoulli.p` is `CompositeImplicitAutograd`.

This is a problem when we're using functionalization on top of python tracing: functionalization will convert `bernoulli_.float` into `bernoulli.p`, but the decomposition for `bernoulli.p` will run before hitting python, and happily convert back into `bernoulli_.float`.

This doesn't seems to come up very often, but I think the right fix is to just make `bernoulli.p` a primitive w.r.t. the Python key too, by making it `CompositeExplicitAutograd`.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix bernoulli_ for functionalization by making it a primitive wrt torch_dispatch"

Context: `bernoulli_.float` is currently a primitive op, but it's functional counterpart `bernoulli.p` is `CompositeImplicitAutograd`.

This is a problem when we're using functionalization on top of python tracing: functionalization will convert `bernoulli_.float` into `bernoulli.p`, but the decomposition for `bernoulli.p` will run before hitting python, and happily convert back into `bernoulli_.float`.

This doesn't seems to come up very often, but I think the right fix is to just make `bernoulli.p` a primitive w.r.t. the Python key too, by making it `CompositeExplicitAutograd`.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix bernoulli_ for functionalization by making it a primitive wrt torch_dispatch"

Context: `bernoulli_.float` is currently a primitive op, but it's functional counterpart `bernoulli.p` is `CompositeImplicitAutograd`.

This is a problem when we're using functionalization on top of python tracing: functionalization will convert `bernoulli_.float` into `bernoulli.p`, but the decomposition for `bernoulli.p` will run before hitting python, and happily convert back into `bernoulli_.float`.

This doesn't seems to come up very often, but I think the right fix is to just make `bernoulli.p` a primitive w.r.t. the Python key too, by making it `CompositeExplicitAutograd`.




[ghstack-poisoned]
